homework/grayscale-image/compression.cpp

```diff
@@ -0,0 +1,58 @@
+#include "compression.hpp"
+#include <algorithm>
+#include <iostream>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(const std::array<std::array<uint8_t, width>, height>& image) {
+	std::vector<std::pair<uint8_t, uint8_t>> compressedImage;
+	compressedImage.reserve(width * height);
+	for (auto verticalIter = image.cbegin(); verticalIter != image.cend(); verticalIter++) {
+		std::vector<uint8_t> elements;
+		elements.reserve(width);
+		std::vector<uint8_t> amounts(width);
+		std::fill(amounts.begin(), amounts.end(), 1);
+		auto amountsIt = amounts.begin();
+		std::unique_copy((*verticalIter).cbegin(),
+				 		 (*verticalIter).cend(),
+						 std::back_inserter(elements),
+						 [&amountsIt](auto current, auto next) mutable {
+							 if (current == next) {
+								 (*amountsIt)++;
+							 } else {
+								 std::advance(amountsIt, 1);
+							 }
+						 return current == next;
+						 });
+		std::transform(elements.cbegin(),
+					   elements.cend(),
+					   amounts.cbegin(),
+					   std::back_inserter(compressedImage),
+					   [](auto first, auto second) -> std::pair<uint8_t, uint8_t> {
+						   return {first, second};
+					   });
+	}
+	compressedImage.shrink_to_fit();
+	return compressedImage;
+}
+
+std::array<std::array<uint8_t, width>, height> decompressGrayscale(const std::vector<std::pair<uint8_t, uint8_t>>& compressedImage) {
+	std::array<std::array<uint8_t, width>, height> image;
+	auto imageRowIt = image.begin();
+	auto imageElemIt = (*imageRowIt).begin();
+	for (auto compressedImageIt = compressedImage.cbegin(); compressedImageIt != compressedImage.cend(); compressedImageIt++) {
+		imageElemIt = std::fill_n(imageElemIt, (*compressedImageIt).second, (*compressedImageIt).first);
+		if (imageElemIt == (*imageRowIt).end()) {
+			imageRowIt++;
+			imageElemIt = (*imageRowIt).begin();
+		}
+	}

```

Comment:
Use iterator returned from std::fill_n. Vector is conigiuous so all elemnts are in the same palce. When you handle one row, the next element is the begining of the second row. So you just need to keep track an iterator and pass it,

`it = std::fill_n(it, N, val);`

Vector2D: {{x, x, x}, {y, y,y}, {z,z,z}
Is sotred in emory like this:
0  1  2  3  4  5  6  7  8
[x][x][x][y][y][y][z][z][z]